syntax = "proto3";

package track;

service Track {
  rpc CreateTrack(CreateTrackRequest) returns (TrackResponse) {}
  rpc GetTrack(TrackRequest) returns (TrackResponse) {}
  rpc GetTracks(TracksRequest) returns (TracksResponse) {}
  rpc GetTracksByIds(TracksByIdsRequest) returns (TracksResponse) {}
  rpc SearchTracks(SearchRequest) returns (TracksResponse) {}

  rpc CreateAuthor(CreateAuthorRequest) returns (AuthorResponse) {}
}

message Author {
  string id = 1;
  string name = 2;
  string genres = 3;
}

message CreateTrackRequest {
  string title = 1;
  repeated string author_ids = 2;
  uint32 duration = 3;
  string source = 4;
  string thumbnail = 5;
}

message TrackRequest {
  string id = 1;
}

message TrackResponse {
  string id = 1;
  string title = 2;
  repeated Author authors = 3;
  uint32 duration = 4;
  string source = 5;
  string thumbnail = 6;
}

message TracksRequest {
  uint32 limit = 1;
  uint32 offset = 2;
}

message TracksByIdsRequest {
  repeated string ids = 1;
}

message TracksResponse {
  repeated TrackResponse tracks = 1;
}

message SearchRequest {
  string query = 1;
  uint32 limit = 2;
  uint32 offset = 3;
}

message CreateAuthorRequest {
  string name = 1;
  string genres = 2;
}

message AuthorResponse {
  string id = 1;
  string name = 2;
  string genres = 3;
}